lib/memcached/memcached.rb

```diff
@@ -585,18 +589,37 @@ def stats(subcommand = nil)
     raise e
   end
 
+  # Causes get / cas operations to return NOT_FOUND instead of raising NotFound
+  def without_exceptions(&block)
+    prev = @raise_on_errors
+    @raise_on_errors = false
+    yield
+  ensure
+    @raise_on_errors = prev
+  end
+
   ### Operations helpers
 
   private
 
   # Checks the return code from Rlibmemcached against the exception list. Raises the corresponding exception if the return code is not Memcached::Success or Memcached::ActionQueued. Accepts an integer return code and an optional key, for exception messages.
+  # When @raise_on_errors is false, `NOT_FOUND` is returned instead of raising an exception when the key is not found.
   def check_return_code(ret, key = nil) #:doc:
-    if ret == 0 # Lib::MEMCACHED_SUCCESS
-    elsif ret == 32 # Lib::MEMCACHED_BUFFERED
-    elsif ret == 16
-      raise NotFound, "NOTFOUND".freeze, @show_backtraces, cause: nil # Lib::MEMCACHED_NOTFOUND
-    elsif ret == 14
-      raise NotStored, "NOTSTORED".freeze, @show_backtraces, cause: nil # Lib::MEMCACHED_NOTSTORED
+    case ret
+    when 0  # Lib::MEMCACHED_SUCCESS
+    when 32 # Lib::MEMCACHED_BUFFERED
+    when 16
+      if @raise_on_errors
+        raise NotFound, "NOTFOUND".freeze, @show_backtraces, cause: nil # Lib::MEMCACHED_NOTFOUND
+      else
+        NOT_FOUND
+      end
+    when 14
+      if @raise_on_errors
+        raise NotStored, "NOTSTORED".freeze, @show_backtraces, cause: nil # Lib::MEMCACHED_NOTSTORED
+      else
+        NOT_FOUND
```

Comment:
If we're going to try and handle all normal cases, we should return `NOT_STORED` here so that callers can distinguish between these cases, and then also handle 12 -> `DATA_EXISTS`, which can be returned from `cas` if the data has been changed since reading it.